### PR TITLE
SLING-10127 optionally emit converted content packages to dedicated folder

### PR DESCRIPTION
**Pull request by kwin**
improve readme

### CODE REVIEW

**Comment by karlpauls** [comment]
@kwin - I'm not sure we should do this change without providing for an option to not write out the artifact. Could we maybe have that cli switch be telling us if we want to deploy the artifact or not? Right now, you are making drop content configurable but at the same time make it so that the artifact is deploy regardless. I think it would be better if that could be configured as well.

As a side note, I think this was intended to be used together with the csv that is generated by:
https://github.com/apache/sling-org-apache-sling-feature-cpconverter/blob/master/src/main/java/org/apache/sling/feature/cpconverter/vltpkg/DefaultPackagesEventsEmitter.java

I guess the idea was: package of type content will not be changed, hence, not deployed - but listed in the csv. Maybe there should be a better way to handle that all together...

**Comment by kwin** [comment]
@karlpauls I asked in https://issues.apache.org/jira/browse/SLING-10127?focusedCommentId=17311422&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-17311422 for input but haven't received any.
Where do you propose to write the modified content-packages which are not referenced in the feature model? Have a new mandatory CLI parameter?

**Comment by karlpauls** [comment]
@kwin, right, sorry - anyways, I think it makes sense to make the drop content selectable but to write the content package no matter what seems unnecessary. I would rather make that configurable too (either via a separate switch or by overloading the drop switch). I'm fine with deploying it to the maven repo - just would make it so that this only happens if drop content is false or is specifically requested (some way).

**Comment by kwin** [comment]
> but to write the content package no matter what seems unnecessary.
I still fail to see use cases, where the transformed content package is not used (even of type "content"). Can you give an example? As outlined in the JIRA some pieces of content packages might be put into the feature model (i.e. ACLs). Installing those afterwards again is unnecessary and dangerous!

**Comment by karlpauls** [comment]
@kwin, I don't disagree that it might be a mistake to ignore "content" content packages - however, that is what this option is specifically for namely, to drop content. I agree it is a good idea to make it configurable but I would suggest we should not generate artifacts for content that is to be dropped. 

From a purely conceptional POV, if the resulting feature model doesn't reference the artifacts - why should they be "there" in the resulting maven repo? 

Now, clearly, sometimes the conceptional POV isn't all that important - hence, as I said, I think it would be ok to add another option (or overload the drop option) that makes the converter preserve the artifacts despite being dropped, but it should be a choice. 

In that case, I guess I have a slight preference to make that a separate location to store only the dropped packages (which one could still configure to be the same as the other artifacts) but I'm not feeling to strongly about it.

**Comment by kwin** [comment]
@karlpauls `dropContent` is only an internal flag and can imho be renamed to reflect that this is only about not referencing it in the feature model. What matters is IMHO the CLI option which is named `--include-content-type-packages`. I am fine with outputting the artifacts not refenced in the feature model in a dedicated output folder, but I still don't see the need to two CLI options or more than the boolean toggle `--include-content-type-packages`. Again, `dropContent` was never exposed as option but always set to true!

**Comment by karlpauls** [comment]
@kwin, right. So the current behavior is to never reference "content" content packages and never write them. I think we should as a first step expose that as an option. You want to go further and change it to be always write the artifacts and only make the reference optional. What do users do that want the current way of working? We would force them to change what they do - which I would like to avoid.

**Comment by kwin** [comment]
bq. I think we should as a first step expose that as an option.

That is where we don't agree. I don't see a good use case for that (even if it the current status quo). Can you give one?

bq. You want to go further and change it to be always write the artifacts

For me this is not going "further", but probably related to the first point.

bq. What do users do that want the current way of working?

Again, I don't see any use case for the current (errorneous) way of working.

bq. We would force them to change what they do - which I would like to avoid

No, just outputting a non-referenced artifact does not do any harm. It all depends on whether you use it afterwards or not.

**Comment by karlpauls** [comment]
> That is where we don't agree. I don't see a good use case for that (even if it the current status quo). Can you give one?

I'm not sure how to make this any clearer: I do not _disagree_. 

> For me this is not going "further", but probably related to the first point.

It is clearly something new. Right now, the converter doesn't output "content" content packages - with your change it will.

> Again, I don't see any use case for the current (errorneous) way of working.

I agree. Nevertheless, it is how it is currently used.

> No, just outputting a non-referenced artifact does not do any harm. It all depends on whether you use it afterwards or not.

It is putting (potentially lots and big) content packages on disc where they are not needed. 

In summary, you want to change the current behavior - whatever usage there is currently, has to adapt. I don't understand why you want to insist that everybody that wants the old way has to now wrap the output in something that deletes the content package manually that are not used in the first place. 

I understand you think it is wrong to not use them - that I can agree with but why can't we just be BC in the default case and make it an option to include the content packages?

**Comment by kwin** [comment]
I exposed now three options for the `contenttypepackagepolicy` in https://github.com/apache/sling-org-apache-sling-feature-cpconverter/pull/70/commits/9674b0a3f684f1fca98d7bb7e23868e5631f8af0: `DROP` (deprecated, default behaviour), `REFERENCE` and `PUT_IN_DEDICATED_FOLDER`. WDYT?

File: src/main/java/org/apache/sling/feature/cpconverter/artifacts/LocalMavenRepositoryArtifactsDeployer.java [hunk 1]
```diff
@@ -99,4 +92,17 @@ public void deploy(@NotNull ArtifactWriter artifactWriter, @NotNull ArtifactId i
         }
     }
 
+    public static String getNameFromArtifactId(ArtifactId id) {
+        StringBuilder nameBuilder = new StringBuilder()
```
Comment (cziegeler) [inline src/main/java/org/apache/sling/feature/cpconverter/artifacts/LocalMavenRepositoryArtifactsDeployer.java:96]: Why not use ArtifactId.toMvnName() ?

Comment (kwin) [reply src/main/java/org/apache/sling/feature/cpconverter/artifacts/LocalMavenRepositoryArtifactsDeployer.java:96]: I just extracted the existing code into a new method, but even better to reuse the one from `ArtifactId`.

Comment (kwin) [reply src/main/java/org/apache/sling/feature/cpconverter/artifacts/LocalMavenRepositoryArtifactsDeployer.java]: Done in https://github.com/apache/sling-org-apache-sling-feature-cpconverter/pull/70/commits/e569b904545f7b58e81e1242ed2bd87eb58a4f55.

**Review by karlpauls — APPROVED** [review]
Thanks @kwin!

Could you add the new cli options to the README? Otherwise, it looks good to me.

**Pull request merged**